Mat 317

stuartarchibald opened the pull request:
This PR is fairly heavy going I'm afraid. MAT-317 is to increase the coverage of the JNI layer, which as we know is horribly complicated to test. I feel the outcome of this code is that we now have a large number of tests covering a lot of behaviour, which is good, however it emphasises that perhaps the design of the JVMManager class is wrong. The static nature of the methods in the JVMManager (for sound design reasons at the time) make it very hard to test as the underlying JNI calls have to be mocked opposed to mocking the behaviour of the JVMManager methods via class extension and overriding virtuals.

This PR:
- Adds a load of unit tests for the JNI layer (largely terminals and data binding).
- Sorts out problems with windows symbol exports.
- Removes duplicated binding code.
- Fixes a number of erroneous `delete` calls.
- Makes methods that are not implemented throw opposed to just run debug statements (done with view that there's work waiting to make it to mainline that handles exceptions).
- Adds more aggressive checking of JNI calls and state.

I realise that the jshim test code is not that eloquent and has some duplication. However, as a redesign is on the books, I think the effort of neater testing with an mock JVMManager would be best done at that time.

BB Multi-Pass: [http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/662](http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/662)
Coverage: librdag 92.4%, jshim 71.0% (massive improvement in jshim)
Java coverage: 82% (unchanged)

Commit:
[MAT-317] JOGScalar JNI layer test

Commit:
[MAT-317] move bindings to their own test.

Commit:
[MAT-317] cover debug statements

Commit:
[MAT-317] ogmatrix coverage, start of testing

[MAT-317] ogmatrix coverage, end to end testing

Commit:
[MAT-317] getIntFromVoidJMethod coverage

Commit:
[MAT-317] JOGDiag<T> tests

Commit:
[MAT-317] JOG<T>Sparse tests

Commit:
[MAT-317] JOG<T>Sparse complete test

Commit:
[MAT-317] apply windows patch, dllexports now present

Commit:
[MAT-317] add more coverage of JVMManager

Commit:
[MAT-317] check rest of jvm manager

Commit:
[MAT-317] started checking dispatch

Commit:
[MAT-317] patch windows export table

[MAT-317] investigating osx hang

Commit:
[MAT-317] dispatch on scalars

Commit:
[MAT-317] dispatch on diag

Commit:
[MAT-317] dispatch on sparse

Commit:
[MAT-317] 2complex16AoA scalars

Commit:
[MAT-317] 2complex16AoA diag and sparse

stuartarchibald commented:
I rewrote history as the commit log was ridiculous, new BB Mulit-Pass:
[http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/663](http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/663)

gmarkall commented on src/jshim/jdispatch.cc:
```diff
@@ -147,7 +146,7 @@ void DispatchToComplex16ArrayOfArrays::visit(librdag::OGScalar<complex16> const
 void DispatchToComplex16ArrayOfArrays::visit(librdag::OGScalar<real16> const SUPPRESS_UNUSED  *thing)
 {
-    throw std::runtime_error("DispatchToComplex16ArrayOfArrays::visit(librdag::OGScalar<real16>)");
+    throw std::runtime_error("DispatchToComplex16ArrayOfArrays::visit(librdag::OGScalar<int>)");
```
isn't this in `DispatchToComplex16ArrayOfArrays::visit(librdag::OGScalar<real16>` ?

gmarkall commented on src/jshim/test/check_jbindings.cc:
```diff
@@ -0,0 +1,253 @@
+{
+    class Fake_JavaVM_bad_env: public Fake_JavaVM
+    {
+      virtual jint AttachCurrentThread(void SUPPRESS_UNUSED **penv, void SUPPRESS_UNUSED *args) override
```
penv is not unused

gmarkall commented on src/jshim/test/check_jbindings.cc:
```diff
@@ -0,0 +1,253 @@
+    {
+        return this->_env;
+    }
+    virtual jint AttachCurrentThread(void SUPPRESS_UNUSED **penv, void SUPPRESS_UNUSED *args)
```
penv is not unused

gmarkall commented on src/jshim/test/check_jbindings.cc:
```diff
@@ -0,0 +1,253 @@
+{
+    class Fake_JavaVM_bad_env: public Fake_JavaVM
+    {
+      virtual jint AttachCurrentThread(void SUPPRESS_UNUSED **penv, void SUPPRESS_UNUSED *args) override
```
penv is not unused

gmarkall commented on src/jshim/test/check_jterminals.cc:
```diff
@@ -0,0 +1,499 @@
+      _value = value;
+    }
+    // fake object call returns obj
+    virtual jobject CallObjectMethod(jobject SUPPRESS_UNUSED obj, jmethodID SUPPRESS_UNUSED methodID, ...) override
```
obj not unused

gmarkall commented on src/jshim/test/check_jterminals.cc:
```diff
@@ -0,0 +1,499 @@
+    virtual jint CallIntMethod(jobject SUPPRESS_UNUSED obj, jmethodID SUPPRESS_UNUSED methodID, ...) {
+      return _rows;
+    }
+    virtual jobject CallObjectMethod(jobject SUPPRESS_UNUSED obj, jmethodID SUPPRESS_UNUSED methodID, ...) override
```
obj is not unused

gmarkall commented on src/jshim/test/check_jterminals.cc:
```diff
@@ -0,0 +1,499 @@
+    virtual jint CallIntMethod(jobject SUPPRESS_UNUSED obj, jmethodID SUPPRESS_UNUSED methodID, ...) {
+      return this->_rows;
+    }
+    virtual jobject CallObjectMethod(jobject SUPPRESS_UNUSED obj, jmethodID SUPPRESS_UNUSED methodID, ...) override
```
obj is not unused

gmarkall commented:
This looks excellent, and looks like a thorough test of everything that can go right/wrong in the tested parts of jshim. I only see a couple of minor issues and discussion points:
- The `Fake_JavaVM` and `Fake_JNIEnv` are replicated across each jshim test. Is there a reason for this, or can they be combined in a single location?
- `getEnv` is not a real JNI method, and looks like a convenience method for setting up tests. I note that it wasn't actually added as part of this PR, but it looks similar enough to `GetEnv` to trip us (me) up in future - can we change the name to make it more obvious what it is, or just use `GetEnv` instead?
- There are a few places where `SUPPRESS_UNUSED` is annotated on variables that are not actually unused. I started marking all the instances, but then I noticed it was mainly just `penv` and they are probably a copy/paste error. I annotated the other extraneous `SUPPRESS_UNUSED` markings that I saw though.
- Whilst looking at the tests for jbindings/jterminals, I noticed that we check for null being returned by `GetDoubleArrayElements` etc. The JNI spec states that it (and other Get<Type>ArrayElements methods) return null when there is a Java exception pending. However, we immediately throw a C++ exception that will be transformed into a Java one by our outer try/catch. I think the result of this will be that the original Java exception may be getting suppressed (I'm not 100% sure - if 2 exceptions happen in JNI, which one wins? is it the last one?).
- The dispatch functions are throwing `std::runtime_error`, and I think this is a hangover from our original (pre-defining our own exceptions) implementation. I think these can be changed to `convert_error` now.

stuartarchibald commented:
Thanks for reviewing...
- The replication is because I like knowing where things are due to my small mental stack :) I'll merge in a tick.
- Following a long discussion, [MAT-329](http://jira.opengamma.com/browse/MAT-329) is there to deal with the general `env` problems and to make the manager threadsafe in the process. Unfortunately it's going to create a lot of method churn.
- Will remove `SUPPRESS_UNUSED` where needed.
- I rather suspect its random undefined behaviour so I'll correct such that checkEx() is called post call.
- Yeah will fix.

Commit:
[MAT-317] swap run time errors for convert_errors

Commit:
[MAT-317] patch up tests and code so that checkEx is correct and used in place of null check.

Commit:
[MAT-317] factor out Fake_{JVM,JNIEnv} to header file for easy enclusion and deduping code.

stuartarchibald commented:
I think I've addressed 1,3,4,5 if you wouldn't mind taking a look please? And added a ticket for 2 as that's a few days work!
New BB Multi-pass: [http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/666](http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/666)

gmarkall commented on src/jshim/test/check_jterminals.cc:
```diff
@@ -0,0 +1,436 @@
+{
+  class Fake_JavaVM_bad_attach: public Fake_JavaVM
+  {
+    virtual jint AttachCurrentThread(void SUPPRESS_UNUSED **penv, void SUPPRESS_UNUSED *args) override
```
penv is not unused here

Commit:
[MAT-317] respond to PR comment

stuartarchibald commented:
Patched [https://github.com/OpenGamma/nyqwk2/pull/68#discussion_r10512423](https://github.com/OpenGamma/nyqwk2/pull/68#discussion_r10512423) in 9a51a81. 
BB Multi-Pass: [http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/667](http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/667)

gmarkall commented:
Apologies for my last comment! This looks great.
